In [1]:
import pandas as pd

df = pd.read_csv("messy_sales.csv")

print(df.shape)
display(df.head(10))
print(df.dtypes)

(300, 6)


,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


order_id      int64
date            str
product         str
price       float64
qty           int64
zip           int64
dtype: object


In [2]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [3]:
fill_value = df["price"].median()
print(fill_value)

df["price"] = df["price"].fillna(fill_value)
df.isna().sum()

37.53


order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [4]:
print(df.duplicated().sum())
df = df.drop_duplicates()
print(df.shape)

8
(292, 6)


In [5]:
df["zip"] = df["zip"].astype(str).str.zfill(5)
print(df["zip"].head(10))

0    60614
1    30303
2    10001
3    98101
4    90405
5    30303
6    02134
7    90405
8    60614
9    90210
Name: zip, dtype: str


In [6]:
df["zip"] = df["zip"].astype(str).str.zfill(5)
print(df["zip"].head(10))

0    60614
1    30303
2    10001
3    98101
4    90405
5    30303
6    02134
7    90405
8    60614
9    90210
Name: zip, dtype: str


In [7]:
df["date"] = pd.to_datetime(df["date"], format="mixed")
print(df.dtypes)

order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object


In [8]:
print(df[df["qty"] < 0])

     order_id       date    product  price  qty    zip
202      1140 2026-04-18     webcam  38.69   -5  98101
262      1233 2026-05-09  desk lamp  22.64   -4  10001
297      1025 2026-04-27   keyboard  47.30   -2  02116


In [9]:
df.to_csv("sales_clean.csv", index=False)
print(df.shape)

(292, 6)


## Cleaning Log

- Loaded 300 rows.
- Found 12 missing prices and filled them using the median.
- Removed 8 duplicate rows. 292 rows remained.
- Converted ZIP codes to text and restored leading zeros.
- Converted dates to datetime format.
- Found 3 negative quantities. I kept them because they may represent product returns.
- Final dataset has 292 rows and 6 columns.

A wrong cleaning decision could change the sales totals and lead to incorrect business decisions.

## Initial Observations

- Some prices are missing.
- ZIP codes are stored as numbers, so some leading zeros are missing.
- Dates use different formats.

## AI Use Appendix

Tool used: ChatGPT

Key prompts:
- Help me debug and complete the pandas data-cleaning steps.
- Help me identify missing values, duplicates, ZIP code formatting, and date conversion.